### PR TITLE
[BugFix] Fix Mooncake load ownership in AscendMultiConnector

### LINKED ISSUE #13474: [Bug][P/D]: AscendStore+MooncakeLayerwise MultiConnector: double-write and Prefill IndexError when AscendStore is selected to load
### Your current environment

<details>
<summary>The output of `python collect_env.py`</summary>

```text
OS: Ubuntu 22.04.5 LTS (aarch64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0
Libc version: glibc-2.35
Python version: 3.12.13 (64-bit runtime)
CPU: HiSilicon Kunpeng-920 aarch64, 256 cores

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] pyzmq==27.1.0
[pip3] torch==2.10.0+cpu
[pip3] torch_npu==2.10.0.post2
[pip3] transformers==5.5.4
vLLM Version: 0.23.0
vLLM Ascend Version: 0.1.dev4104+g9a12de829.d20260804 (git sha: 9a12de829, date: 20260804)
mooncake-transfer-engine-npu: 0.3.11.post1

NPU: 910B4 (A2 series), driver/npu-smi 25.5.2
CANN: 9.0.1 (V100R001C10SPC002B220)
```

</details>

### Summary

PD disaggregation with Qwen3-8B (Prefill and Decode both TP=1), fronted by `examples/disaggregated_prefill_v1/load_balance_proxy_layerwise_server_example.py`. Both sides use `MultiConnector` wrapping **`AscendStoreConnector` first and `MooncakeLayerwiseConnector` second** (AscendStore must come first because of the MultiConnector FirstWin mechanism, so that KV Pool hits take precedence on the Decode side). The Decode side sets `consumer_is_to_load: true` and `load_async: true`.

This combination was introduced by PR #7032 (`AscendMultiConnector` was created specifically for it), and the block-count mismatch reported in #7085 was said to be fixed by that PR. However, on the current version, when **AscendStore wins the FirstWin on the Decode side**, we observe two independently reproducible problems:

1. **Double-write**: the losing `MooncakeLayerwiseConnector` still receives real blocks and arms a P2P recv into the same HBM blocks that AscendStore is loading into.
2. **Prefill crash**: when the store hit length is block-aligned, Prefill dies with `IndexError` in `get_transfer_mappings` due to a 73-local vs 72-remote block mismatch.

### Configuration

Decode-side `kv_transfer_config`:

```json
{
  "kv_connector": "MultiConnector",
  "kv_role": "kv_consumer",
  "kv_load_failure_policy": "recompute",
  "kv_connector_extra_config": {
    "connectors": [
      {
        "kv_connector": "AscendStoreConnector",
        "kv_role": "kv_consumer",
        "kv_connector_extra_config": {
          "backend": "mooncake",
          "lookup_rpc_port": "1",
          "load_async": true,
          "consumer_is_to_load": true
        }
      },
      {
        "kv_connector": "MooncakeLayerwiseConnector",
        "kv_role": "kv_consumer",
        "kv_port": "30200",
        "kv_connector_extra_config": {
          "prefill": {"dp_size": 1, "tp_size": 1},
          "decode": {"dp_size": 1, "tp_size": 1}
        }
      }
    ]
  }
}
```

Prefill side uses the symmetric `kv_producer` configuration with `kv_port: 30000` and `lookup_rpc_port: "0"`.

### Problem 1: Losing MooncakeLayerwise still receives real blocks — both connectors write the same HBM

Because of the special case in `AscendMultiConnector.update_state_after_alloc` (`isinstance(c, MooncakeLayerwiseConnector)`), after AscendStore wins the FirstWin, `MooncakeLayerwiseConnector` **still receives the full real block list and `num_external_tokens`**. As a result:

- Both Decode-side workers write into the same HBM blocks: the AscendStore worker issues a pool `get` into them, while the MooncakeLayerwise worker registers a P2P recv so Prefill pushes into the very same blocks.
- Decode-side MooncakeLayerwise still calls back the proxy's `/v1/metaserver` with `remote_cached_tokens=0`, which makes **Prefill re-send the request's entire KV via P2P from offset 0** — pure waste, racing the pool load on the same memory.
- `MultiConnector.get_finished` unions the two connectors' `done_recving` sets, so the request is released for decoding as soon as **either** load finishes, while the other connector may still be writing those blocks.

Key Decode-side logs (prompt = 9248 tokens, store hit = 9216):

```text
MultiConnector.get_num_new_matched_tokens req=... connector=0:AscendStoreConnector num_computed_tokens=0 -> toks=9216 load_async=True
MooncakeLayerwise.get_num_new_matched_tokens req=... do_remote_prefill=True num_computed_tokens=0 prompt_len=9248 -> matched=9248 load_async=True
MultiConnector FirstWin req=... winner=0(AscendStoreConnector) toks=9216 load_async=True
MultiConnector.update_state_after_alloc req=... chosen_connector=0(AscendStoreConnector) num_external_tokens=9216 blocks_per_group=[72]
  req=... connector=1:MooncakeLayerwiseConnector receives REAL blocks via MooncakeLayerwise special-case (NOT the FirstWin winner)
AscendStore.update_state_after_alloc req=... can_load=True num_external_tokens=9216 vllm_cached=0 kvpool_cached=9216 blocks_per_group=[72]
MooncakeLayerwise.update_state_after_alloc req=... CONSUMER recv-armed: num_external_tokens=9216 blocks_per_group=[72] remote_cached_tokens=0
AscendStoreWorker.start_load_kv req=... queued ASYNC pool load: token_len=9216 vllm_cached=0 kvpool_cached=9216
MooncakeLayerwiseWorker.start_load_kv CONSUMER req=... registered for P2P recv: blocks_per_group=[72] (Prefill P2P data will be written into these same HBM blocks)
```

### Problem 2: Block-aligned store hit → Prefill `IndexError` in `get_transfer_mappings`

When AscendStore hits 9216 tokens (72×128, exactly block-aligned, so the "full hit minus one token" rule does not trigger):

- Decode allocates only **72 blocks** for the external tokens, and the MooncakeLayerwise recv path advertises those 72 blocks to the metaserver as `remote_block_ids`.
- Prefill plans its send for the full prompt (9248 tokens = **73 blocks**), because the advertised `remote_cached_tokens` is 0.
- While building the transfer mappings (73 local blocks → 72 remote blocks), Prefill indexes past the remote list and EngineCore dies — the whole Prefill pod goes down.

For contrast: when the store misses (MooncakeLayerwise wins FirstWin with the full prompt length), Decode allocates 73 blocks, both sides agree, and no crash occurs — i.e. the crash is an edge condition introduced precisely when AscendStore wins FirstWin.

<details>
<summary>Full traceback (Prefill side)</summary>

```text
ERROR [core.py:1197] EngineCore encountered a fatal error.
Traceback (most recent call last):
  ...
  File "/vllm-workspace/vllm-ascend/vllm_ascend/worker/model_runner_v1.py", line 2258, in execute_model
    self.maybe_get_kv_connector_output(
  File "/vllm-workspace/vllm/vllm/v1/worker/kv_connector_model_runner_mixin.py", line 95, in _get_kv_connector_output
    kv_connector.start_load_kv(get_forward_context())
  File "/vllm-workspace/vllm/vllm/distributed/kv_transfer/kv_connector/v1/multi_connector.py", line 291, in start_load_kv
    c.start_load_kv(forward_context, **kwargs)
  File "/vllm-workspace/vllm-ascend/vllm_ascend/distributed/kv_transfer/kv_p2p/mooncake_layerwise_connector.py", line 1675, in start_load_kv
    single_group_transfer_mappings = self._get_kv_split_metadata(req_meta, req_idx, req_id, i)
  File "/vllm-workspace/vllm-ascend/vllm_ascend/distributed/kv_transfer/kv_p2p/mooncake_layerwise_connector.py", line 1575, in _get_kv_split_metadata
    transfer_mappings = get_transfer_mappings(
  File "/vllm-workspace/vllm-ascend/vllm_ascend/distributed/kv_transfer/utils/utils.py", line 294, in get_transfer_mappings
    d_block_id = d_block_ids[d_block_idx]
IndexError: list index out of range
```

</details>

`SchedulerOutput` dumped by `dump_input.py` right before the crash shows the block-count mismatch directly:

```text
remote_block_ids: [[74, 75, ..., 145]]     # advertised by Decode: 72 blocks
remote_cached_tokens: 0
local block_ids=([1, 2, ..., 72, 74],)     # Prefill side: 73 blocks
prefix_cache_stats=PrefixCacheStats(queries=9248, hits=9216)
```

### Reproduction steps

1. Start Prefill (`kv_producer`) / Decode (`kv_consumer`) / proxy / `mooncake_master` with the config above.
2. Send one long-prompt request (e.g. 9248 tokens) and confirm its KV is stored in the KV Pool.
3. Invalidate the Decode-side local HBM prefix cache (we used `POST /reset_prefix_cache`; any "HBM miss but store hit" scenario is equivalent).
4. Send the same prompt again:
   - **Problem 1** (double write + redundant Prefill re-send) reproduces every time.
   - **Problem 2** (Prefill crash) reproduces whenever the store hit length is block-aligned.

### Expected behavior

When AscendStore wins FirstWin:

- `MooncakeLayerwiseConnector` should **not** receive the request's real blocks
- it should **not** arm a P2P recv
- Prefill should **not** be triggered to re-send the full KV
- Decode should load from the KV Pool only

### Questions

1. Is `MultiConnector[AscendStoreConnector, MooncakeLayerwiseConnector]` with AscendStore ordered first plus `consumer_is_to_load` a supported usage? PR #7032 suggests the combination is intended to work, but the behaviors above show the "store wins FirstWin" path is not handled.
2. Is the `MooncakeLayerwiseConnector` special case in `AscendMultiConnector.update_state_after_alloc` (handing real blocks to the losing connector) intentional? If so, how are the double write and the redundant Prefill re-send supposed to be avoided?
3. If this usage is explicitly unsupported, it would be safer to fail fast with a clear error/warning on this path instead of silently double-writing or crashing at runtime.

### PR DESCRIPTION
**Pull request by Futuresxy**
This PR aims to fix #13474.

When `MultiConnector[AscendStoreConnector, MooncakeLayerwiseConnector]`
uses FirstWin and AscendStore wins a Decode-side load, the Ascend-specific
`update_state_after_alloc` override currently gives the losing Mooncake
connector both the real blocks and the winner's `num_external_tokens`.
Mooncake consequently arms a second receive into the same HBM blocks, calls
the metaserver, and can trigger a redundant full P-to-D transfer. For a
block-aligned store hit this also creates a local/remote block-count mismatch
and Prefill can fail in `get_transfer_mappings` with `IndexError`.

This patch separates block visibility from load ownership:

- the FirstWin winner receives real blocks and the external token count;
- a losing Mooncake connector keeps real blocks, which its producer-side push
  path still needs, but receives zero external tokens;
- Mooncake remote-prefill handling is side-effect-free for zero external
  tokens: it does not read block IDs, enqueue a receive, or call the
  metaserver;
- `AscendMultiConnector` identifies the Mooncake child through
  `MultiConnector`'s aligned child configuration instead of eagerly importing
  the optional Mooncake implementation.

The producer-side `do_remote_decode` path remains unchanged and continues to
derive its send metadata from the real blocks. This also matches current vLLM
MultiConnector's ownership contract, where non-winners receive real blocks
with zero external tokens.

### Does this PR introduce _any_ user-facing change?

Yes. With AscendStore ordered before MooncakeLayerwise, a Decode-side KV Pool
hit now loads only through AscendStore instead of also starting a redundant
Mooncake receive. There is no API or configuration change. Importing
`AscendMultiConnector` for non-Mooncake configurations also no longer requires
the optional `mooncake` package.

### How was this patch tested?

- `pytest -q tests/ut/kv_offload/test_ascend_multi_connector.py tests/ut/kv_offload/test_mooncake_layerwise_connector.py`
  - `45 passed, 25 warnings in 2.13s`
- Isolated import check with `importlib.util.find_spec("mooncake") is None`;
  `AscendMultiConnector` imported successfully.
- `ruff check`, `ruff format --check`, codespell, typos, and the repository's
  Python policy hooks on all four changed files.
- Qwen3-4B service baseline on one 910C using the verified vLLM main commit.
- MooncakeLayerwise 1P1D smoke test on two 910Cs.
- AscendStore + MooncakeLayerwise MultiConnector 1P1D test on two 910Cs:
  - prompt length 1056, block size 128;
  - first request used the P2P path and stored 1024 tokens / 8 blocks;
  - after resetting Decode's local prefix cache, the second request hit and
    loaded all 8 blocks from AscendStore;
  - Mooncake received zero external tokens and skipped remote prefill;
  - metaserver callback count remained `1 -> 1`;
  - both requests completed and no `IndexError` occurred.

The aggregate local pre-commit command could not execute its gitleaks hook
because the repository helper downloaded an x86-64 binary on the aarch64
host; the remaining hooks above passed. The latest GitHub E2E workflow (`#17569`), including gitleaks,
pre-commit, and mypy, passed.

- vLLM version: v0.26.0
- vLLM main: https://github.com/vllm-project/vllm/commit/58d3918e3ea0a544ffedadad2ba84559e9c51d8f

### CODE REVIEW

File: vllm_ascend/distributed/kv_transfer/ascend_multi_connector.py [hunk 1]
```diff
@@ -33,9 +33,13 @@ def update_state_after_alloc(self, request: "Request", blocks: "KVCacheBlocks",
...
+            if i == chosen_connector:
                 # Forward call to the chosen connector (if any).
                 c.update_state_after_alloc(request, blocks, num_external_tokens)
+            elif isinstance(c, MooncakeLayerwiseConnector):
+                # Preserve real blocks for Mooncake's producer-side push path,
+                # but do not grant load ownership after it loses FirstWin.
+                c.update_state_after_alloc(request, blocks, 0)
             else:
                 # Call with empty blocks for other connectors.
                 c.update_state_after_alloc(request, empty_blocks, 0)
```
[the remark this answers is not in the document]
Comment (Futuresxy) [reply vllm_ascend/distributed/kv_transfer/ascend_multi_connector.py:43]: Addressed in 4d5e0b8. I removed the top-level Mooncake implementation import. Rather than importing inside `update_state_after_alloc`, the code uses the child `KVTransferConfig` list that `MultiConnector` already keeps aligned with `_connectors`. This preserves lazy loading for non-Mooncake users and avoids repeated imports or masking an internal Mooncake `ImportError` on the scheduler hot path. I also added an import-isolation regression test; the focused suite passes (45 tests), and an isolated environment with `find_spec("mooncake") is None` can now import `AscendMultiConnector`.
